packages/phishing-controller/src/PhishingDetector.ts

```diff
@@ -0,0 +1,177 @@
+import { distance } from 'fastest-levenshtein';
+
+import {
+  domainPartsToDomain,
+  domainPartsToFuzzyForm,
+  domainToParts,
+  getDefaultPhishingDetectorConfig,
+  matchPartsAgainstList,
+  processConfigs,
+} from './utils';
+
+export type LegacyPhishingDetectorList = {
+  whitelist?: string[];
+  blacklist?: string[];
+} & FuzzyTolerance;
+
+export type PhishingDetectorList = {
+  allowlist?: string[];
+  blocklist?: string[];
+  name?: string;
+  version?: string | number;
+} & FuzzyTolerance;
+
+export type FuzzyTolerance =
+  | {
+      tolerance?: number;
+      fuzzylist: string[];
+    }
+  | {
+      tolerance?: never;
+      fuzzylist?: never;
+    };
+
+export type PhishingDetectorOptions =
+  | LegacyPhishingDetectorList
+  | PhishingDetectorList[];
+
+export type PhishingDetectorConfiguration = {
+  name?: string;
+  version?: number | string;
+  allowlist: string[][];
+  blocklist: string[][];
+  fuzzylist: string[][];
+  tolerance: number;
+};
+
+export type PhishingDetectorResult = {
+  name?: string;
+  version?: string;
+  result: boolean;
+  match?: string;
+  type: 'all' | 'fuzzy' | 'blocklist' | 'allowlist' | 'blacklist' | 'whitelist';
+};
+
+export class PhishingDetector {
+  #configs: PhishingDetectorConfiguration[];
+
+  #legacyConfig: boolean;
+
+  /**
+   * Construct a phishing detector, which can check whether origins are known
+   * to be malicious or similar to common phishing targets.
+   *
+   * A list of configurations is accepted. Each origin checked is processed
+   * using each configuration in sequence, so the order defines which
+   * configurations take precedence.
+   *
+   * @param opts - Phishing detection options
+   */
+  constructor(opts: PhishingDetectorOptions) {
+    // recommended configuration
+    if (Array.isArray(opts)) {
+      this.#configs = processConfigs(opts);
+      this.#legacyConfig = false;
+      // legacy configuration
+    } else {
+      this.#configs = [
+        getDefaultPhishingDetectorConfig({
+          allowlist: opts.whitelist,
+          blocklist: opts.blacklist,
+          fuzzylist: opts.fuzzylist,
+          tolerance: opts.tolerance,
+        }),
+      ];
+      this.#legacyConfig = true;
+    }
+  }
+
+  /**
+   * Check if a domain is known to be malicious or similar to a common phishing
+   * target.
+   *
+   * @param domain - The domain to check.
+   * @returns The result of the check.
+   */
+  check(domain: string): PhishingDetectorResult {
+    const result = this.#check(domain);
+
+    if (this.#legacyConfig) {
+      let legacyType = result.type as PhishingDetectorResult['type'];
+      if (legacyType === 'allowlist') {
+        legacyType = 'whitelist';
+      } else if (legacyType === 'blocklist') {
+        legacyType = 'blacklist';
+      }
+      return {
+        match: result.match,
+        result: result.result,
+        type: legacyType,
+      };
+    }
+    return result;
+  }
+
+  #check(domain: string): PhishingDetectorResult {
```

Comment:
This can most likely be merged into the public `check` method since all the public method does is handle the differences with the legacy result type, so that we don't have two methods with the same name without a clear responsibility distinction.

I'm unsure about doing it in this PR though, since there are many things to review already